File: src/message_queue.cpp
#include <iostream>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <thread>
#include <chrono>
#include <optional>

// A thread-safe message queue
template <typename T>
class MessageQueue {
public:
    // Push a message into the queue
    void send(T msg) {
        std::lock_guard<std::mutex> lock(mtx_);
        queue_.push(std::move(msg));
        cond_var_.notify_one();
    }

    // Try to receive a message with a timeout
    std::optional<T> recv_timeout(std::chrono::milliseconds timeout) {
        std::unique_lock<std::mutex> lock(mtx_);

        if (!cond_var_.wait_for(lock, timeout, [this] { return !queue_.empty(); })) {
            return {};  // Timeout: return an empty optional
        }

        T msg = std::move(queue_.front());
        queue_.pop();
        return msg;
    }

private:
    std::queue<T> queue_;             // Queue holding the messages
    std::mutex mtx_;                  // Mutex for thread-safe access to the queue
    std::condition_variable cond_var_; // Condition variable for blocking threads
};

// Function to simulate message sending after a delay
void delayed_send(MessageQueue<std::string>& mq, int delay_seconds, std::string message) {
    std::this_thread::sleep_for(std::chrono::seconds(delay_seconds));
    mq.send(std::move(message));
}

int main() {
    MessageQueue<std::string> mq;

    // Start a thread that sends a message after 2 seconds
    std::thread sender_thread(delayed_send, std::ref(mq), 2, "Hello from the thread!");

    // Try to receive a message with a timeout of 1 second
    auto result = mq.recv_timeout(std::chrono::milliseconds(1000));

    if (result) {
        std::cout << "Received: " << *result << std::endl;
    } else {
        std::cout << "Error: Timeout occurred" << std::endl;
    }

    // Wait for the sender thread to finish
    sender_thread.join();

    return 0;
}

File: src/message_queue.rs
use std::sync::{Arc, Mutex, Condvar};
use std::thread;
use std::time::Duration;

// Define a thread-safe message queue using Mutex and Condvar
struct MessageQueue<T> {
    queue: Mutex<Vec<T>>,         // Mutex-protected queue
    cond_var: Condvar,            // Condition variable to signal when a message is available
}

impl<T> MessageQueue<T> {
    fn new() -> Self {
        MessageQueue {
            queue: Mutex::new(Vec::new()),    // Initialize an empty queue
            cond_var: Condvar::new(),         // Initialize the condition variable
        }
    }

    // Send a message into the queue and notify waiting threads
    fn send(&self, msg: T) {
        let mut queue = self.queue.lock().unwrap();
        queue.push(msg);
        self.cond_var.notify_one();           // Notify one waiting thread that a message is available
    }

    // Try to receive a message with a timeout
    fn recv_timeout(&self, timeout: Duration) -> Result<T, &'static str> {
        let mut queue = self.queue.lock().unwrap();
        
        // If the queue is empty, we will wait for the specified timeout
        let result = self.cond_var.wait_timeout(queue, timeout).unwrap();
        queue = result.0;  // Re-acquire the lock after waiting

        // Check if we received a message or timed out
        if !queue.is_empty() {
            Ok(queue.remove(0)) // Return the message from the front of the queue
        } else {
            Err("Timeout occurred") // Timeout if the queue is still empty
        }
    }
}

fn delayed_send(mq: Arc<MessageQueue<String>>, delay_seconds: u64, message: String) {
    thread::sleep(Duration::from_secs(delay_seconds)); // Simulate delay
    mq.send(message);                                  // Send the message
}

fn main() {
    let mq = Arc::new(MessageQueue::new());

    // Spawn a thread that will send a message after 2 seconds
    let mq_clone = mq.clone();
    thread::spawn(move || {
        delayed_send(mq_clone, 1, "Hello from the thread!".to_string());
    });

    // Try to receive a message with a timeout of 1 second
    match mq.recv_timeout(Duration::from_secs(2)) {
        Ok(msg) => println!("Received: {}", msg),
        Err(err) => println!("Error: {}", err),
    }
}
